### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by cs-suyog-jinturkar**
PMDB: https://pmdb.fortinet.com/ProjectManagement/editprojects.php?id=24273

Authentication mode supported: Basic and HMAC
Action Supported: make an api call
Sample Playbook: Yes

UTC:
1. Successful health check using both auth methods
2. Successful GET, POST, DELETE and PUT API execution using basic and HMAC auth
3. Tested with headers, query param and body as an action inputs

### CODE REVIEW

File: remoteFortisoar/utils.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+def login(config):
+    auth_type = config.get('auth_type')
+    if auth_type == 'Basic':
+        return loginUsingBasicAuth(config)
+    else:
+        return loginUsingHmacAuth(config)
+  
+  
+def loginUsingHmacAuth(config, data=None):
```
Comment (cs-suyog-jinturkar) [inline remoteFortisoar/utils.py]: I will update these fun name in my next PR along with other changes.

File: remoteFortisoar/info.json [hunk 2]
```diff
@@ -0,0 +1,146 @@
+{
+    "name": "remoteFortisoar",
+    "version": "1.0.0",
+    "label": "Remote FortiSOAR",
+    "description": "This connector allows you to run REST API operations on other FortiSOAR environments",
+    "publisher": "Fortinet",
+    "icon_small_name": "",
+    "icon_large_name": "",
+    "cs_approved": false,
+    "category": "",
```
Comment (cs-suyog-jinturkar) [inline remoteFortisoar/info.json]: what should be the category?

File: remoteFortisoar/utils.py (hunk 1 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/utils.py]: done

File: remoteFortisoar/info.json [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                            "value": ""
+                        },
+                        {
+                            "description": "Password used to authenticate FortiSOAR Server",
+                            "title": "Password",
+                            "tooltip": "Password used to authenticate FortiSOAR Server",
+                            "required": true,
+                            "editable": true,
+                            "visible": false,
+                            "type": "text",
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: password

File: remoteFortisoar/info.json [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                            "value": ""
+                        },
+                        {
+                            "description": "Public key used to authenticate FortiSOAR Server",
+                            "title": "Public Key",
+                            "tooltip": "Public key used to authenticate FortiSOAR Server",
+                            "required": true,
+                            "editable": true,
+                            "visible": false,
+                            "type": "textarea",
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json:79]: password

File: remoteFortisoar/info.json [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                            "tooltip": "Private key used authenticate FortiSOAR Server",
+                            "name": "private_key",
+                            "type": "textarea",
+                            "visible": true,
+                            "editable": true,
+                            "required": true,
+                            "value": ""
+                        },
+                        {
+                            "description": "Public key used to authenticate FortiSOAR Server",
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json:62]: can we take public key input before the private key?

File: remoteFortisoar/info.json [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+                            "value": ""
+                        }
+                    ],
+                    "HMAC": [
+                        {
+                            "description": "Private key used to authenticate FortiSOAR Server",
+                            "title": "Private Key",
+                            "tooltip": "Private key used authenticate FortiSOAR Server",
+                            "name": "private_key",
+                            "type": "textarea",
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: password

File: remoteFortisoar/info.json [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                            "required": true,
+                            "value": ""
+                        },
+                        {
+                            "description": "Public key used to authenticate FortiSOAR Server",
+                            "title": "Public Key",
+                            "tooltip": "Public key used to authenticate FortiSOAR Server",
+                            "required": true,
+                            "editable": true,
+                            "visible": false,
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: why visible false?

File: remoteFortisoar/info.json [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+                {
+                    "title": "Endpoint IRI",
+                    "type": "text",
+                    "name": "iri",
+                    "required": true,
+                    "visible": true,
+                    "editable": true
+                },
+                {
+                    "title": "Data",
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: We should provision for request payload, query parameters, and file input, and headers

File: remoteFortisoar/info.json [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                    "name": "data",
+                    "required": false,
+                    "visible": true,
+                    "editable": true
+                }
+            ],
+            "open": false
+        }
+    ],
+    "forked_from": false
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: can we remove this additional keys that get added from byoc?

File: remoteFortisoar/info.json [hunk 10]
```diff
@@ -0,0 +1,145 @@
+{
+    "name": "remoteFortisoar",
+    "version": "1.0.0",
+    "label": "Remote FortiSOAR",
+    "description": "This connector allows you to run REST API operations on other FortiSOAR environments",
+    "publisher": "Fortinet",
+    "icon_small_name": "fortisoar_remote_small.png",
+    "icon_large_name": "fortisoar_remote_large.png",
+    "cs_approved": false,
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: true?

File: remoteFortisoar/info.json (hunk 2 above)
Comment (cs-pooja-singh) [reply remoteFortisoar/info.json]: Utilities

File: remoteFortisoar/operations.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+def make_api_call(config, params, *args, **kwargs):
+    endpoint = params.get('iri')
+    data = params.get('data', None)
+    method = params.get('method')
+    if not endpoint or not method:
+        logger.warning('Got endpoint: {endpoint}\nData: {data}'.format(endpoint=endpoint, data=json.dumps(data)))
+        raise ConnectorError('Missing required input')
+
+    api_response = invoke_rest_endpoint(config, endpoint, method, data)
```
Comment (cs-pooja-singh) [inline remoteFortisoar/operations.py]: allow for query params and file upload also
we will need file for config import.

File: remoteFortisoar/utils.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    else:
+        logger.exception('Invalid authentication type: {0}'.format(auth_type))
+        raise ConnectorError('Invalid authentication type: {0}'.format(auth_type))
+    
+    try:
+        if (auth_type == 'Basic'):
+            response = requests.request(method, full_uri, verify=verify_ssl, json=data, headers=headers)
+        else:
+            if data:
+                response = requests.request(method, full_uri, auth=hmac_auth, verify=verify_ssl, headers=headers, json=data)
```
Comment (cs-pooja-singh) [inline remoteFortisoar/utils.py]: where do we hmacify the payload?

File: remoteFortisoar/playbooks/playbooks.json [hunk 13]
```diff
@@ -0,0 +1,99 @@
+{
```
Comment (cs-pooja-singh) [inline remoteFortisoar/playbooks/playbooks.json:1]: can we remove the .DS_Store above?

File: remoteFortisoar/info.json (hunk 2 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: Done

File: remoteFortisoar/info.json (hunk 10 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: done

File: remoteFortisoar/info.json (hunk 6 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: done

File: remoteFortisoar/info.json (hunk 5 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json:62]: done

File: remoteFortisoar/info.json (hunk 7 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: Marked visible = true

File: remoteFortisoar/info.json (hunk 9 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: Done

File: remoteFortisoar/info.json (hunk 8 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: Created mantis #0883479 to track this requirement

File: remoteFortisoar/operations.py (hunk 11 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/operations.py]: Created mantis #0883479 to track this requirement

File: remoteFortisoar/info.json (hunk 3 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: done

File: remoteFortisoar/playbooks/playbooks.json (hunk 13 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/playbooks/playbooks.json:1]: Done

File: remoteFortisoar/info.json [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+                            "required": true,
+                            "value": ""
+                        },
+                        {
+                            "description": "Password used to authenticate FortiSOAR Server",
+                            "title": "Password",
+                            "tooltip": "Password used to authenticate FortiSOAR Server",
+                            "required": true,
+                            "editable": true,
+                            "visible": false,
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: true. not allowing to configure basic auth currently

File: remoteFortisoar/info.json [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+                                "visible": true,
+                                "editable": true,
+                                "required": false,
+                                "placeholder": "{'id':'1'}"
+                            },
+                            {
+                                "title": "Body",
+                                "name": "body",
+                                "description": "The body (which won't be encoded) to send with the request",
+                                "type": "text",
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: let's put textarea to give more space

File: remoteFortisoar/info.json [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+                                "tooltip": "Body for POST OR PUT methods"
+                            },
+                            {
+                                "title": "Headers",
+                                "name": "headers",
+                                "description": "An dict full of headers to add to the request",
+                                "type": "text",
+                                "visible": true,
+                                "editable": true,
+                                "required": false
```
Comment (cs-pooja-singh) [inline remoteFortisoar/info.json]: a default header of accept: Application/json would be nice to have

File: remoteFortisoar/info.json (hunk 14 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: Done

File: remoteFortisoar/info.json (hunk 15 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: Done

File: remoteFortisoar/info.json (hunk 16 above)
Comment (cs-suyog-jinturkar) [reply remoteFortisoar/info.json]: Done

**Review by cs-pooja-singh — APPROVED** [review]

**Pull request merged**